UNOMI-291 Implement profile matching (section 4.21 of spec)

pmi opened the pull request:
- Extracted property condition creation functions to a separate factory to avoid duplication in fetchers

anatol-sialitski commented on graphql/cxs-impl/src/main/java/org/apache/unomi/graphql/fetchers/SegmentConnectionDataFetcher.java:
```diff
@@ -21,6 +21,7 @@
 import org.apache.unomi.api.conditions.Condition;
 import org.apache.unomi.api.segments.Segment;
 import org.apache.unomi.api.services.DefinitionsService;
+import org.apache.unomi.graphql.condition.ProfileConditionFactory;
```
please to remove unused import

anatol-sialitski commented on graphql/cxs-impl/src/main/java/org/apache/unomi/graphql/fetchers/ProfileConnectionDataFetcher.java:
```diff
@@ -21,6 +21,7 @@
 import org.apache.unomi.api.Profile;
 import org.apache.unomi.api.conditions.Condition;
 import org.apache.unomi.api.services.DefinitionsService;
+import org.apache.unomi.graphql.condition.ProfileConditionFactory;
```
Please to remove unused import

anatol-sialitski commented on graphql/cxs-impl/src/main/java/org/apache/unomi/graphql/fetchers/EventConnectionDataFetcher.java:
```diff
@@ -19,6 +19,7 @@
 
 import org.apache.unomi.api.Event;
 import org.apache.unomi.api.PartialList;
+import org.apache.unomi.graphql.condition.EventConditionFactory;
```
Please to remove unused import

anatol-sialitski commented on graphql/cxs-impl/src/main/java/org/apache/unomi/graphql/types/output/CDPFilterMatch.java:
```diff
@@ -23,12 +23,29 @@
-    public Integer executionTimeMillis;
+    private Long executionTimeMillis;
 
+    public CDPFilterMatch(String name, Boolean matched, Long executionTimeMillis) {
```
Will be better to specify `@GraphQLName("...")` for each parameter

anatol-sialitski commented on graphql/cxs-impl/src/main/java/org/apache/unomi/graphql/condition/ProfileConditionFactory.java:
```diff
@@ -0,0 +1,246 @@
+                rootSubConditions.add(createPropertyCondition("metadata.scope", filterInput.getViewEquals(), definitionsService));
+            }
+
+            if (filterInput.getAndFilters() != null && filterInput.getAndFilters().size() > 0) {
```
Please use `!isEmpty()` instead of `size() > 0` and apply that for the rest of usages

anatol-sialitski commented on graphql/cxs-impl/src/main/java/org/apache/unomi/graphql/condition/ProfileConditionFactory.java:
```diff
@@ -0,0 +1,246 @@
+            }
+
+            if (filterInput.getSegments_contains() != null && filterInput.getSegments_contains().size() > 0) {
+                rootSubConditions.add(createPropertiesCondition("segments", "hasSomeOf", filterInput.getSegments_contains(), definitionsService));
```
You can use the `profileSegmentCondition` condition type

anatol-sialitski commented on graphql/cxs-impl/src/main/java/org/apache/unomi/graphql/condition/ProfileConditionFactory.java:
```diff
@@ -0,0 +1,246 @@
+            }
+
+            if (filterInput.getLists_contains() != null && filterInput.getLists_contains().size() > 0) {
+                rootSubConditions.add(createPropertiesCondition("lists", "hasSomeOf", filterInput.getLists_contains(), definitionsService));
```
You can use the `profileUserListCondition` condition type

anatol-sialitski commented on graphql/cxs-impl/src/main/java/org/apache/unomi/graphql/condition/ProfileConditionFactory.java:
```diff
@@ -0,0 +1,246 @@
+        final Condition rootCondition = createBoolCondition("and", definitionsService);
+        final List<Condition> rootSubConditions = new ArrayList<>();
+
+        if (filterInput.getTopic_equals() != null) {
```
That will not works
Interest is a map, where topic is a key, score is a value, for example:
```
"properties": {
    "interests": {
      "interestName1": 10,
      "interestName2": 5,
      "interestName3": 7
    }
  }
```

anatol-sialitski commented on graphql/cxs-impl/src/main/java/org/apache/unomi/graphql/condition/ProfileConditionFactory.java:
```diff
@@ -0,0 +1,246 @@
+    }
+
+    private Condition createProfileEventsFilterInputCondition(CDPProfileEventsFilterInput filterInput, DefinitionsService definitionsService) {
+        final Condition rootCondition = createBoolCondition("and", definitionsService);
```
The `pastEventCondition` condition should be used for that case

anatol-sialitski commented on graphql/cxs-impl/src/main/java/org/apache/unomi/graphql/condition/ConditionFactory.java:
```diff
@@ -0,0 +1,88 @@
+        this.entityName = entityName;
+    }
+
+    public static ProfileConditionFactory profile() {
```
That is not thread safe

pmi committed:
UNOMI-291 Implement profile matching (section 4.21 of spec)

- Extracted property condition creation functions to a separate factory to avoid duplication in fetchers
